plugins/out_stackdriver/stackdriver.c

```diff
@@ -277,6 +277,114 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+static int process_local_resource_id(const void *data, size_t bytes,
+                                     struct flb_stackdriver *ctx, char *type)
+{
+    int i;
+    int len;
+    int ret_code = -1;
+    int flag = 0;
+    size_t off = 0;
+    msgpack_object k;
+    msgpack_object v;
+    msgpack_object root;
+    msgpack_object_map map;
+    msgpack_unpacked result;
+    char *local_resource_id;
+    char *ptr;
+    const char delim[] = ".";
+
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS && !flag) {
+        root = result.data;
+        flag = 0;
+
+        if (root.type != MSGPACK_OBJECT_ARRAY ||
+            root.via.array.size != 2 ||
+            root.via.array.ptr[1].type != MSGPACK_OBJECT_MAP) {
+            flb_plg_warn(ctx->ins, "unexpected record format");
+            continue;
+        }
+
+        map = root.via.array.ptr[1].via.map;
+        len = strlen(LOCAL_RESOURCE_ID_KEY);
+        for (i = 0; i < map.size; i++) {
+            k = map.ptr[i].key;
+            v = map.ptr[i].val;
+
+            if (k.type == MSGPACK_OBJECT_STR && 
+                strncmp(k.via.str.ptr, LOCAL_RESOURCE_ID_KEY, len) == 0) {
+                
+                ret_code = 0;
+                local_resource_id = malloc(v.via.str.size+1);
+                strncpy(local_resource_id, v.via.str.ptr, v.via.str.size);
+                local_resource_id[v.via.str.size] = '\0';
+
+
+                ptr = strtok(local_resource_id, delim);
+                /* Skip the prefix of tag */
+                ptr = strtok(NULL, delim);
+
+                if (strncmp(type, "k8s_container", strlen("k8s_container")) == 0) {
+                    while (ptr != NULL)
+                    {
+                        /* Follow the order of fields in local_resource_id */
+                        if (!ctx->namespace_name) {
+                            ctx->namespace_name = flb_sds_create(ptr);
+                        }
+                        else if (!ctx->pod_name) {
+                            ctx->pod_name = flb_sds_create(ptr);
+                        } 
+                        else if (!ctx->container_name) {
+                            ctx->container_name = flb_sds_create(ptr);
+                        }
+
+                        ptr = strtok(NULL, delim);
+                    }
+
+                    if (!ctx->namespace_name || !ctx->pod_name || !ctx->container_name) {
```

Comment:
I suggest to destroy all resources and return -1 directly instead of returning the code at the end of this function and destroy them in stackdriver_format function.
It's clear that which resources are created and they should be free once the object fails to construct. The remaining code in this function are mostly not related to 'k8s_container' type.

Reply:
Done. Let the function return -1 once the condition is not met.